rilldata/rill
# feat: Add self serve billing management UI

Adds a self serve billing page under organisation settings. Some local setup is needed to get this working.

Contact me on slack for stripe and orb keys to be added to `.env`. Once `.env` is updated run `rill devtool start cloud --refresh-dotenv=false` to start cloud.

## Trial subscription
To test trail subscription there is a helper command to advance time artifically. (might get merged into `devtool` subcommand in `rill`) : https://github.com/rilldata/rill-helpers?tab=readme-ov-file#trail-subscription
`--org` is optional and will default to the org selected using `rill org switch`
- After starting a fresh subscription, run `./rill-helpers sub advance-time --org <org_name> --time <today+2>` to get the `Ending soon` email. This is mandatory to keep things consistant.
- Run `./rill-helpers sub advance-time --org <org_name>` next to end the trial.
- Run `./rill-helpers sub advance-time --org <org_name>` next to end the grace period as well.

## Team plan
To test team plans webhooks forwarder is needed for stripe and orb.
- Signup to ngrok: https://ngrok.com
- Follow steps to download and authenticate: https://dashboard.ngrok.com/get-started/setup/macos
- Create a domain: https://dashboard.ngrok.com/domains
- Start ngrok with above domain: `ngrok http --url=<domain name> 8080`
- Using url `https://<domain>/payment/webhook`,  add webhook to stripe:  https://dashboard.stripe.com/test/webhooks.(select all events while creating this)
- Copy over the secret to `RILL_ADMIN_STRIPE_WEBHOOK_SECRET`. Reveal signing secret on the webhook page to get this.
- Using `https://<domain>/billing/webhook` as url, add webhook to orb: https://app.withorb.com/webhooks (Make sure to be on test plan).
- Copy over the secret to `RILL_ADMIN_ORB_WEBHOOK_SECRET`. Get the signing secret from webhook page.
- Restart cloud server if already started. Use `rill devtool start cloud --refresh-dotenv=false` to avoid overwriting the keys.

To test a cancelled team plan expiring (needs some familiarity with orb),
- Make sure the plan is cancelled.
- Go to orb test account : https://app.withorb.com/customers.
- Look for your org's customer using org name.
- "Unschedule cancellation" and cancel again by selecting "cancel immediately"
- Run `./rill-helpers sub advance-time --org <org_name>` to reflect things in rill.

To test a failed payment,
- Go to stripe using org's billing settings section.
- Add a card that will fail payment: https://docs.stripe.com/testing#declined-payments
- Go to https://app.withorb.com/invoices/new.
- Find the customer and create a new invoice.
- Make sure to select "Payment terms" as "Due on issue".
- After creating make sure to manually issue it.
- Rill cloud should now catch the failed payment.
- Advance time to test grace period passing: `./rill-helpers sub advance-time --org <org_name>`
